Add simple ollama chat example

matthewhaynesonline opened the pull request:
Fixes #191. 

I used the code examples from https://github.com/64bit/async-openai/issues/173 as a starting point and tried to simplify them to match the existing chat example.

64bit commented on async-openai/README.md:
```diff
@@ -59,6 +59,11 @@ $Env:OPENAI_API_KEY='sk-...'
 - Visit [examples](https://github.com/64bit/async-openai/tree/main/examples) directory on how to use `async-openai`.
 - Visit [docs.rs/async-openai](https://docs.rs/async-openai) for docs.
 
+### Local LLM Usage
```
This is not required. Ollama example has its own README

64bit commented on examples/ollama-chat/README.md:
```diff
@@ -0,0 +1,18 @@
+## Prerequisites
```
Please consider adding fully contained example - in this case if there's dependency on ollama - please consider creating docker-compose file to bring it up in single command. 

I'd prefer to have quick way to run an example, instead having them additional steps - which a user may already do on their own

matthewhaynesonline commented:
Hi, I've updated my PR with your feedback, though the Docker setup has some caveats I didn't have elegant solutions for. 

1. **Build context:** since Docker restricts where you can copy files from to the current directory branch (parents / siblings are not allowed), I adjusted the example to pull the `async_openai` crate from crates.io, not the parent dir. 
   1. The only way I could think to use the crate from the parent project dir was to place the `docker-compose.yml` / `Dockerfile` in the project root, but that would no longer make the example self contained.
2. **First time Ollama model pull:** Ollama requires you to pull a model before you can use it and, understandably, the Ollama docker image doesn't come preloaded with any models. As such, a custom entrypoint is use to pull the model on startup. A healthcheck is then used to wait to start the Rust container until after the Ollama container is fully ready.

What do you think?

64bit commented:
Thanks for update, I just tried it and looks good! 

I think `Dockerfile` and `ollam_chat` containers  in docker-compose are not required. My initial reason for docker-compose was to bring up Ollama and then `cargo run` example which would points to "localhost:<ollama-port>".

Your entrypoint stuff to download model looks good to me.

So user has to do two steps:
```
# bring ollama up with model
docker compose up -d

# once model is downloaded and ollama is up
cargo run 
```

64bit commented on examples/ollama-chat/Dockerfile:
```diff
@@ -0,0 +1,9 @@
+FROM rust:1.85 AS builder
```
Not required

64bit commented on examples/ollama-chat/docker-compose.yml:
```diff
@@ -0,0 +1,40 @@
+services:
+  web:
+    container_name: ollama_chat
+    build: .
+    command: ollama-chat
+    restart: on-failure:3
+    environment:
+      API_BASE: "http://ollama:11434/v1"
+      API_KEY: "ollama"
+      MODEL: "llama3.2:1b"
+    depends_on:
+      ollama:
+        condition: service_healthy
+
```
not required

64bit commented on examples/ollama-chat/ollama_entrypoint.sh:
```diff
@@ -0,0 +1,16 @@
+sleep 5
+
+echo "Retrieving model $MODEL..."
+ollama pull $MODEL
```
Nice

64bit commented on examples/ollama-chat/src/main.rs:
```diff
@@ -0,0 +1,62 @@
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn Error>> {
+    let api_base = env::var("API_BASE").expect("API_BASE is not set in the environment.");
```
This can be hard coded to what docker compose will bring up 

so `localhost:11434`

matthewhaynesonline commented:
@64bit Sounds good! I updated my branch with your feedback to run the Rust code natively and rebased / cleaned up the commit history off latest upstream main.

64bit reviewed (approved):
Looking good, thank your for your contributions!
